clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -968,13 +968,31 @@ private void transitionTo(State target) {
     }
 
     private void transitionTo(State target, RuntimeException error) {
+        transitionTo(target, error, false);
```

Comment:
> I think changing the default behavior to not throw can cause issues in some calls:
> 
> 1. TransactionManager.InitProducerIdHandler#handleResponse on line 1303 - lastError is explicitly set to null (which shouldn't be done at all, as transitionTo already does that if the state transition is valid), which will clear the latest error. I think to make this work, that lastError = null should be removed from line 1303

I've removed that line. I am not sure why it's there, but `git blame` says it's been there since 2017 😬 

> 2. This is a call chain where we transition on direct user action, shouldn't this be throwing? KafkaProducer.send -> KafkaProducer.doSend -> maybeTransitionToErrorState -> transitionToAbortableError -> transitionTo

Since `maybeTransitionToErrorState` is being called from inside a `catch` block, if we did throw an exception, it would mask the root issue (`ApiException`), right?

> 3. In TransactionManager.TxnOffsetCommitHandler#handleResponse, there are multiple
> 
> ```
> abortableError(...);
> break;
> ```
> 
> blocks. If abortableError does not throw on invalid state transition anymore, the txn commit will be retried, even when in a failed state, which doesn't seem correct.

After the `abortableError` call, if the state transition was invalid, then `currentState` would be `FATAL_ERROR`. That has the same effect as the last two branches of the `if` statement that call the `fatalError` method, right?

Would simply skipping the reenqueue on fatal errors be sufficient?

```
if (result.isCompleted()) {
    pendingTxnOffsetCommits.clear();
} else if (pendingTxnOffsetCommits.isEmpty()) {
    result.done();
} else {
    if (!hasFatalError()) {
        // Retry the commits which failed with a retriable error
        reenqueue();
    }
}
```

I don't yet understand why we'd want to re-enqueue after those calls to `fatalError` anyway 🤔

Reply:
> > This is a call chain where we transition on direct user action, shouldn't this be throwing? KafkaProducer.send -> KafkaProducer.doSend -> maybeTransitionToErrorState -> transitionToAbortableError -> transitionTo
> 
> Since maybeTransitionToErrorState is being called from inside a catch block, if we did throw an exception, it would mask the root issue (ApiException), right?

I think masking the ApiException is better than silently transitioning into fatal state - if transitionToAbortableError tries going into abortable state, that ApiException is probably something that the calling code can handle, and try to recover by aborting. If we still throw that exception, but in reality the internal state is fatal already, that is a violation of the API, isn't it?

> After the abortableError call, if the state transition was invalid, then currentState would be FATAL_ERROR. That has the same effect as the last two branches of the if statement that call the fatalError method, right?
> 
> Would simply skipping the reenqueue on fatal errors be sufficient?

I think you are right, and the last 2 branches of that if do the same. At a later point, Sender.runOnce will handle the fatal state and will stop sending, but maybe it would be cleaner to not reenqueue at this point anymore, yes.

Reply:
For the last point, it is probably out of scope, and needs a more generic approach for all transactional reenqueue calls

Reply:
> I think masking the ApiException is better than silently transitioning into fatal state - if transitionToAbortableError tries going into abortable state, that ApiException is probably something that the calling code can handle, and try to recover by aborting. If we still throw that exception, but in reality the internal state is fatal already, that is a violation of the API, isn't it?

I've updated the code to throw the exception in this case, as you'd suggested.